registry/client/auth/session.go

```diff
@@ -32,71 +34,24 @@ type CredentialStore interface {
 
 // NewAuthorizer creates an authorizer which can handle multiple authentication
 // schemes. The handlers are tried in order, the higher priority authentication
-// methods should be first.
-func NewAuthorizer(transport http.RoundTripper, handlers ...AuthenticationHandler) RequestModifier {
-	return &tokenAuthorizer{
-		challenges: map[string]map[string]authorizationChallenge{},
+// methods should be first. The challengeMap holds a list of challenges for
+// a given root API endpoint (for example "https://registry-1.docker.io/v2/").
+func NewAuthorizer(challengeMap map[string][]Challenge, handlers ...AuthenticationHandler) transport.RequestModifier {
```

Comment:
Possibly not the most straight forward

```go
challengeMap := map[string][]auth.Challenge{}
resp, _ := http.Get(endpointStr)
_, err := auth.Check(resp, "")
if err != nil {
	if challengeErr, ok := err.(*auth.ErrAuthorizationChallenge); ok {
		challengeMap[endpointStr] = challengeErr.Challenges
	}
}
```

I would still expect this to be wrapped in some sort of ping method though. Either a method to ping multiple endpoints or single endpoint.